### PR TITLE
Within and between distance API methods for the DissimilarityMatrix object

### PR DESCRIPTION
**Pull request by wasade**
This pull request adds in `.within` and `.between` methods to the `DissimilarityMatrix` object, and expressing the data as a `pd.DataFrame`. These methods provide a standard way for obtaining within and between group distances, a common operation for beta-diversity derived distance matrices.

Please complete the following checklist:

* [x] I have read the guidelines in [CONTRIBUTING.md](https://github.com/biocore/scikit-bio/blob/master/CONTRIBUTING.md).

* [x] I have documented all public-facing changes in [CHANGELOG.md](https://github.com/biocore/scikit-bio/blob/master/CHANGELOG.md).

* [ ] **This pull request includes code, documentation, or other content derived from external source(s).** If this is the case, ensure the external source's license is compatible with scikit-bio's [license](https://github.com/biocore/scikit-bio/blob/master/COPYING.txt). Include the license in the `licenses` directory and add a comment in the code giving proper attribution. Ensure any other requirements set forth by the license and/or author are satisfied. **It is your responsibility to disclose code, documentation, or other content derived from external source(s).** If you have questions about whether something can be included in the project or how to give proper attribution, include those questions in your pull request and a reviewer will assist you.

* [x] **This pull request does not include code, documentation, or other content derived from external source(s).**

**Note:** [REVIEWING.md](https://github.com/biocore/scikit-bio/blob/master/REVIEWING.md) may also be helpful to see some of the things code reviewers will be verifying when reviewing your pull request.

### CODE REVIEW

**Comment by wasade** [comment]
The test failures appear unrelated to this PR with the exception of a Sphinx [warning](https://travis-ci.org/biocore/scikit-bio/jobs/573924648#L1546) that I don't understand. I took a look back at the code for instances of "from" in the `between` method (using `from_` in code, and nothing obvious stood out.

File: skbio/stats/distance/_base.py [hunk 1]
```diff
@@ -366,6 +366,187 @@ def filter(self, ids, strict=True):
...
+            missing = list(ids - set(self._id_index))
+            raise MissingIDError("At least one ID (e.g., '%s') was not "
+                                 "found." % missing[0])
+
+        i_order = self._stable_order(ids)
+        return self._subset_to_dataframe(i_order, i_order)
+
+    @experimental(as_of="0.5.4")
+    def between(self, from_, to_, allow_overlap=False):
+        """Obtain all the distances among the set of IDs
```
Comment (gwarmstrong) [inline skbio/stats/distance/_base.py]: Doesn't this function 'Obtain all the distances from `from_` to `to_`' or less precisely, 'Obtain all the distances between the sets of IDS'?

File: skbio/stats/distance/_base.py [hunk 2]
```diff
@@ -366,6 +366,187 @@ def filter(self, ids, strict=True):
...
+        -------
+        np.array, dtype=int
+            The corresponding index values
+        """
+        # establish a stable ordering, and cache both the positions and label
+        # of each ID in the stable order.
+        id_order = sorted((self._id_index[i] for i in ids))
+        return np.array(id_order, dtype=int)
+
+    @experimental(as_of="0.5.4")
```
Comment (gwarmstrong) [inline skbio/stats/distance/_base.py]: Should this version be upped since the most recent release, which has already been released, is 0.5.5?

File: skbio/stats/distance/_base.py [hunk 3]
```diff
@@ -366,6 +366,187 @@ def filter(self, ids, strict=True):
...
+
+        if ids - set(self._id_index):
+            missing = list(ids - set(self._id_index))
+            raise MissingIDError("At least one ID (e.g., '%s') was not "
+                                 "found." % missing[0])
+
+        i_order = self._stable_order(ids)
+        return self._subset_to_dataframe(i_order, i_order)
+
+    @experimental(as_of="0.5.4")
```
Comment (gwarmstrong) [inline skbio/stats/distance/_base.py]: Same as above

File: skbio/stats/distance/tests/test_base.py [hunk 4]
```diff
@@ -57,6 +64,110 @@ def setUp(self):
...
+                            ['a', 'c', 4.2],
+                            ['c', 'a', 4.2],
+                            ['c', 'c', 0.0]],
+                           columns=['i', 'j', 'value'])
+
+        # NOTE: order was changed from ['a', 'c'] to ['c', 'a']
+        # but the output order in exp is consistent with
+        # test_within
+        obs = self.dm_3x3.within(['c', 'a'])
+        pdt.assert_frame_equal(obs, exp)
```
Comment (gwarmstrong) [inline skbio/stats/distance/tests/test_base.py]: Can you also include:
```
        obs = self.dm_3x3.within(['a', 'c'])
        pdt.assert_frame_equal(obs, exp)
```
In this test, similar to `test_between_order_stability`, just to assure permutation stability within the test.

File: skbio/stats/distance/_base.py (hunk 2 above)
Comment (wasade) [reply skbio/stats/distance/_base.py]: Ah, good catch :)

File: skbio/stats/distance/tests/test_base.py [hunk 5]
```diff
@@ -57,6 +64,112 @@ def setUp(self):
...
+
+        # varying the order of the "i" values, result remains consistent
+        # with the test_between result
+        obs = self.dm_5x5.between(['d', 'b'], ['a', 'c', 'e'])
+        pdt.assert_frame_equal(obs, exp)
+
+        # varying the order of the "j" values, result remains consistent
+        # with the test_between result
+        obs = self.dm_5x5.between(['b', 'd'], ['a', 'e', 'c'])
+        pdt.assert_frame_equal(obs, exp)
```
Comment (fedarko) [inline skbio/stats/distance/tests/test_base.py]: like 99% sure this wouldn't be a problem, but it *might* be worth adding another test where both the i and j values are "out-of-order"? e.g.

```python
obs = self.dm_5x5.between(['d', 'b'], ['a', 'e', 'c'])
pdt.assert_frame_equal(obs, exp)
```

(this could be done with minimal effort)

File: skbio/stats/distance/tests/test_base.py [hunk 6]
```diff
@@ -57,6 +64,112 @@ def setUp(self):
...
+        exp = pd.DataFrame([['b', 'a', 5.],
+                            ['b', 'd', 7.],
+                            ['b', 'e', 8.],
+                            ['d', 'a', 4.],
+                            ['d', 'd', 0.],
+                            ['d', 'e', 7.]],
+                           columns=['i', 'j', 'value'])
+
+        obs = self.dm_5x5._subset_to_dataframe([1, 3], [0, 3, 4])
+        pdt.assert_frame_equal(obs, exp)
```
Comment (fedarko) [inline skbio/stats/distance/tests/test_base.py]: Is passing in an empty np.array for i and/or j to `_subset_to_dataframe()` a thing that could conceivably happen? If so, would be worth adding tests for that case (I imagine this would just result in an empty output).

If this isn't a thing that should ever happen, then it would be worth adding a quick check to the start of `_subset_to_dataframe()` that checks to make sure both `i_indices` and `j_indices` have at least one element (and raises an error if not).

File: skbio/stats/distance/_base.py [hunk 7]
```diff
@@ -366,6 +366,187 @@ def filter(self, ids, strict=True):
...
+            The IDs to establish a stable ordering for.
+
+        Returns
+        -------
+        np.array, dtype=int
+            The corresponding index values
+        """
+        # establish a stable ordering, and cache both the positions and label
+        # of each ID in the stable order.
+        id_order = sorted((self._id_index[i] for i in ids))
```
[the remark this answers is not in the document]
Comment (wasade) [reply skbio/stats/distance/_base.py]: It's a generator expression, so just avoids unrolling the items prior to executing with `sorted`

File: skbio/stats/distance/_base.py [hunk 8]
```diff
@@ -366,6 +366,187 @@ def filter(self, ids, strict=True):
...
+        4  B  B    0.0
+        5  B  C    1.0
+        6  C  A    2.0
+        7  C  B    1.0
+        8  C  C    0.0
+        """
+        ids = set(ids)
+
+        if ids - set(self._id_index):
+            missing = list(ids - set(self._id_index))
```
[the remark this answers is not in the document]
Comment (wasade) [reply skbio/stats/distance/_base.py]: I'm inclined to disagree on these items. For 1, this is using an operator on a primitive data type defined by the language itself, and is to the best of my knowledge PEP8 compliant. But, I'd be happy to revise if you insist. 

For 2, the proposal incurs additional overhead (list unrolling) in all cases, whereas the present code incurs the overhead only in the exceptional case.

File: skbio/stats/distance/_base.py [hunk 9]
```diff
@@ -366,6 +366,187 @@ def filter(self, ids, strict=True):
...
+        2  A  E    4.0
+        3  B  C    1.0
+        4  B  D    2.0
+        5  B  E    3.0
+        """
+        from_ = set(from_)
+        to_ = set(to_)
+
+        if (from_ | to_) - set(self._id_index):
+            missing = list((from_ | to_) - set(self._id_index))
```
[the remark this answers is not in the document]
Comment (wasade) [reply skbio/stats/distance/_base.py]: I feel that the present statements are more succinct and easier to understand than `from_.union(to_).difference(set(self._id_index))`, or variations where many variables are needed. 

@thermokarst, when you have a moment, would it be possible to weigh in as a third party for resolving this?

File: skbio/stats/distance/_base.py [hunk 10]
```diff
@@ -366,6 +366,187 @@ def filter(self, ids, strict=True):
...
+            raise MissingIDError("At least one ID (e.g., '%s') was not "
+                                 "found." % missing[0])
+
+        if not allow_overlap and (from_ & to_):
+            missing = list(from_ & to_)
```
[the remark this answers is not in the document]
Comment (wasade) [reply skbio/stats/distance/_base.py]: Ah, good catch on 2. I disagree on 1 :) so hopefully we can get a consensus from a third party for how to proceed

File: skbio/stats/distance/_base.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (fedarko) [reply skbio/stats/distance/_base.py]: That's fair (and sorry if these comments come across as bikeshedding). I understand where you're coming from with point 2.

How does the following sound? The main idiom that I'm not a fan of throughout this PR is writing out the set operations twice, since that's (albeit very slightly) inefficient and typo-prone. The following solution avoids this, while only doing list unrolling in the exceptional case.
```python
missing_ids = ids - set(self._id_index)
if missing_ids:
    missing_id_list = list(missing_ids)
    raise MissingIDError(...)
```

File: skbio/stats/distance/_base.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (fedarko) [reply skbio/stats/distance/_base.py]: Oh, ok! Sorry, I hadn't seen that in practice before. No need to remove the parens then.

File: skbio/stats/distance/_base.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (fedarko) [reply skbio/stats/distance/_base.py]: Sorry, I should've been more clear—by "more explicit" I don't mean using the set operation names (e.g. `.union()`, I agree that's clunky), I meant something up-front like

```python
missing_ids = (from_ | to_) - set(self._id_index)
if missing_ids:
    missing_ids_list = list(missing_ids)
    ...
```

This way, the set operation is only done once which is 1) slightly more efficient and 2) easier to update if we'd want to change the way missing IDs are computed in the future.

File: skbio/stats/distance/_base.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (wasade) [reply skbio/stats/distance/_base.py]: oh, I see, it's actually an implicit generator expression. Crazy!

```python
>>> sum(i for i in [1,2,3])
6
>>> def foo(x):
...   print(x)
...
>>> foo(i for i in [1,2,3])
<generator object <genexpr> at 0x1029b4ca8>
>>>
```

File: skbio/stats/distance/_base.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (wasade) [reply skbio/stats/distance/_base.py]: I'll meet you half way on (1), commit coming soon :)

File: skbio/stats/distance/_base.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (wasade) [reply skbio/stats/distance/_base.py]: @thermokarst, nm, have a new proposed stylistic solution here so I think all good

File: skbio/stats/distance/tests/test_base.py (hunk 6 above)
Comment (wasade) [reply skbio/stats/distance/tests/test_base.py]: Good catch on the edge case, it does get slightly weird

File: skbio/stats/distance/_base.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (wasade) [reply skbio/stats/distance/_base.py]: @fedarko, I think we hit on the same idea :) that's essentially what just came in on the latest commits

File: skbio/stats/distance/_base.py [hunk 11]
```diff
@@ -495,50 +491,60 @@ def between(self, from_, to_, allow_overlap=False):
...
-            missing = list((from_ | to_) - set(self._id_index))
+        all_ids = from_ | to_
+        not_present = all_ids - set(self._id_index)
+        if not_present:
             raise MissingIDError("At least one ID (e.g., '%s') was not "
-                                 "found." % missing[0])
+                                 "found." % not_present.pop())
 
+        overlapping = from_ & to_
         if not allow_overlap and (from_ & to_):
```
Comment (fedarko) [inline skbio/stats/distance/_base.py]: extremely minor, but we can replace this line with

```python
if not allow_overlap and overlapping:
```

File: skbio/stats/distance/_base.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (fedarko) [reply skbio/stats/distance/_base.py]: @wasade Awesome, I'm glad we could come to an agreement! The new changes look great. Thanks for all your work on adding this to scikit-bio :)

**Comment by wasade** [comment]
I think this is good to merge?

**Comment by fedarko** [comment]
Sorry for the delay in responding. Everything looks good to merge (IMO, at least).
